src/ecs.rs

```diff
@@ -0,0 +1,692 @@
+            match field.key() {
+                "name" => {
+                    let unquoted_reference =
+                        quoted_string::to_content::<TraceQuoteSpec>(field.value())
+                            .map_err(|_| fmt::Error)?;
+                    reference_name = Some(unquoted_reference.to_string());
+                }
+                "accept" => {
+                    plugin_field_set.accept =
+                        str::parse::<f64>(field.value()).map_err(|_| fmt::Error)?;
+                }
+                "restrict" => {
+                    plugin_field_set.restrict =
+                        str::parse::<f64>(field.value()).map_err(|_| fmt::Error)?;
+                }
+                "unknown" => {
+                    plugin_field_set.unknown =
+                        str::parse::<f64>(field.value()).map_err(|_| fmt::Error)?;
+                }
+                "score" => {
+                    plugin_field_set.score =
+                        str::parse::<f64>(field.value()).map_err(|_| fmt::Error)?;
+                }
+                _ => {}
+            }
+        }
+
+        // These messages may be sent up to twice for each plugin, once for the request phase and once for
+        // the response phase. If there is a message for both phases, the response phase information will
+        // overwrite the previous request phase information. This may ultimately prove to be a problematic
+        // design. It's optimizing for verbosity and total log transfer, but it's lossy by discarding values
+        // that are superceded by decisions in later phases. Since decisions values from earlier phases are
+        // carried to the next phase, it should still always report the decision values that contributed to
+        // the final decision.
+        if let Some(reference_name) = reference_name {
+            let mut bulwark = ecs_event.bulwark.clone().unwrap_or_default();
+            let mut plugins = bulwark.plugins.clone().unwrap_or_default();
+            plugins.insert(
+                reference_name,
+                serde_json::to_value(plugin_field_set).map_err(|_| fmt::Error)?,
+            );
+            bulwark.plugins = Some(plugins);
+            ecs_event.bulwark = Some(bulwark);
+        } else {
+            return Err(fmt::Error);
+        }
+
+        Ok(())
+    }
+
+    /// Parses `"combine decision"` messages emitted for request and response phases which combine individual
+    /// plugin decisions into an ensemble decision.
+    ///
+    /// The decision outcome based on the configured decision thresholds and the combined set of all tags emitted
+    /// by the plugins is also parsed from these messages.
+    fn parse_combine_decision_event(event: &Event, ecs_event: &mut EcsEvent) -> fmt::Result {
+        // If the service is in observe-only mode, don't report events as denied
+        // Fields aren't sorted, scanning because we can't binary search
+        let observe_only = event
```

Comment:
Yes, the entire ECS module is one big hack. The `tracing-forest` crate is being used to keep multi-line logging for non-ECS log formats from interleaving across threads, making the logs unreadable. I probably could have used `tracing-tree` for ECS instead since its limitations are far less applicable there, but I think a lot of what makes the ECS formatter be a hack would apply in that scenario too.

I may still end up exploring that switch because I suspect there's some overhead that `tracing-forest` introduces. Since ECS logging doesn't take advantage of what differentiates the two crates, any overhead is just waste.